<?xml version="1.0"?>
<robot xmlns:xacro="http://www.ros.org/wiki/xacro">


  <xacro:property name="whisker_mesh_file" value="file://$(find rm3_gazebo)/urdf/meshes/whisker.dae"/>
  <xacro:property name="whisker_array_mesh_file" value="file://$(find rm3_gazebo)/urdf/meshes/whisker_array.dae"/>

  <xacro:property name="array_step" value="0.0315"/>
  <!-- <xacro:property name="grid_step" value="0.04"/> -->
  <xacro:property name="flex_base_thickness" value="0.0"/>
  <!-- <xacro:property name="whisker_length" value="0.0575"/> -->
  <xacro:property name="whisker_length" value="0.150"/>
  <xacro:property name="whisker_radius" value="0.002"/>
  <xacro:property name="whisker_body_mass" value="0.015"/>
  <xacro:property name="base_link_mass" value="0.1"/>
  <xacro:property name="whisker_array_mass" value="2.75"/>
  <xacro:property name="whisker_revolute_effort_limit" value="10"/>
  <xacro:property name="whisker_revolute_vel_limit" value="${pi/2}"/>

  <xacro:property name="whisker_x" value="-0.0345"/>
  <xacro:property name="whisker_x_bias" value="0.075"/>
  <xacro:property name="whisker_y" value="0.001"/>
  <xacro:property name="whisker_z" value="0.02"/>



 <!-- WHISKER DEFITION -->
  <xacro:macro name="whisker" params="name id parent *origin">
      <!-- Magnet around x axis is connected to whisker array-->

        <joint name="${name}_magnet_x_${id}" type="revolute">
        <xacro:insert_block name="origin"/>
        <parent link="${parent}"/>
        <child link="${name}_magnet_x_link_${id}"/>
        <axis xyz="1 0 0" />
        <limit
          lower="${-pi/2}"
          upper="${pi/2}"
          effort="${whisker_revolute_effort_limit}"
          velocity="${whisker_revolute_vel_limit}" />
        <dynamics
          damping="0.1"
          friction="0.1" />
        </joint>

        <link name="${name}_magnet_x_link_${id}">
        <inertial>
        <mass value="0.0005" />
        <inertia ixx="1.0e-4" ixy="0" ixz="0" iyy="1.0e-4" iyz="0" izz="1.0e-4" />
        </inertial>
        </link>

        <!-- Magnet around y axis is connected to whisker array-->

        <joint name="${name}_magnet_y_${id}" type="revolute">
        <parent link="${name}_magnet_x_link_${id}"/>
        <child link="${name}_magnet_y_link_${id}"/>
        <axis xyz="0 1 0" />
        <limit
          lower="${-pi/2}"
          upper="${pi/2}"
          effort="${whisker_revolute_effort_limit}"
          velocity="${whisker_revolute_vel_limit}" />
        <dynamics
          damping="0.1"
          friction="0.1" />
        </joint>

        <link name="${name}_magnet_y_link_${id}">
        <inertial>
        <mass value="0.0005" />
        <inertia ixx="1.0e-4" ixy="0" ixz="0" iyy="1.0e-4" iyz="0" izz="1.0e-4" />
        </inertial>
        </link>

      <!-- Whisker fixed on magnet y-->
      <joint name="${name}_whisker_${id}" type="fixed">
          <origin xyz="0 0 0" rpy="0 0 0" />
          <!-- <parent link="${name}_magnet_y_link_${id}"/> -->
           <parent link="${name}_magnet_y_link_${id}"/>
          <child link="${name}_whisker_link_${id}"/>
      </joint>

      <link name="${name}_whisker_link_${id}">
          <inertial>
            <mass
              value="${whisker_body_mass}" />
            <inertia
              ixx="${1/12 * whisker_body_mass * (3* whisker_radius ** 2 + whisker_length **2 )}" ixy="0" ixz="0"
              iyy="${1/12 * whisker_body_mass * (3* whisker_radius ** 2 + whisker_length **2 )}" iyz="0"
              izz="${1/2 * whisker_body_mass * whisker_radius ** 2}" />
          </inertial>
          <visual>
              <geometry>
                <mesh filename="${whisker_mesh_file}" scale="1 1 1"/>
              </geometry>
          </visual>

          <collision>
               <geometry>
                 <mesh filename="${whisker_mesh_file}" scale="1 1 1"/>
               </geometry>
          </collision>
      </link>
  </xacro:macro>


 <!-- WHISKER ARRAY DEFITION -->
  <xacro:macro name="whisker_array" params="name *origin">
      <joint name="${name}_whisker_array" type="fixed">
          <xacro:insert_block name="origin"/>
          <parent link="chassis"/>
          <child link="${name}_whisker_array_link"/>
      </joint>

      <link name="${name}_whisker_array_link">
        <inertial>
          <xacro:insert_block name="origin"/>
          <mass value="${whisker_array_mass}" />
          <inertia ixx="0.0012" ixy="0.0" ixz="0.0"
                 iyy="0.00079" iyz="0.0"
                 izz="0.0079" />
          </inertial>

          <visual>
              <geometry>
                <mesh filename="${whisker_array_mesh_file}" scale="1 1 1"/>
              </geometry>
          </visual>

        <collision>
               <geometry>
                 <mesh filename="${whisker_array_mesh_file}" scale="1 1 1"/>
               </geometry>
        </collision>
      </link>

      <gazebo reference="${name}_whisker_array_link">
        <mu1 value="1"/>
        <mu2 value="1"/>
        <kp value="100000.0" />
        <kd value="1.0" />
        <material>Gazebo/DarkGrey</material>
      </gazebo>

      <!-- whisker #1 -->
      <xacro:whisker name="${name}" id = "1" parent= "${name}_whisker_array_link" >
        <origin xyz="${whisker_x + -3*whisker_x_bias} ${whisker_y} ${whisker_z}"
                rpy="0 0 0" />
      </xacro:whisker>
      <!-- whisker #2 -->
      <xacro:whisker name="${name}" id = "2" parent= "${name}_whisker_array_link" >
        <origin xyz="${whisker_x + -2*whisker_x_bias} ${whisker_y} ${whisker_z}"
                rpy="0 0 0" />
      </xacro:whisker>
      <!-- whisker #3 -->
      <xacro:whisker name="${name}" id = "3" parent= "${name}_whisker_array_link" >
        <origin xyz="${whisker_x + -whisker_x_bias} ${whisker_y} ${whisker_z}"
                rpy="0 0 0" />
      </xacro:whisker>
      <!-- whisker #4 -->
      <xacro:whisker name="${name}" id = "4" parent= "${name}_whisker_array_link" >
        <origin xyz="${whisker_x + -0*whisker_x_bias} ${whisker_y} ${whisker_z}"
                rpy="0 0 0" />
      </xacro:whisker>
      <!-- whisker #5 -->
      <xacro:whisker name="${name}" id = "5" parent= "${name}_whisker_array_link" >
        <origin xyz="${whisker_x + whisker_x_bias} ${whisker_y} ${whisker_z}"
                rpy="0 0 0" />
      </xacro:whisker>
      <!-- whisker #6 -->
      <xacro:whisker name="${name}" id = "6" parent= "${name}_whisker_array_link" >
        <origin xyz="${whisker_x + 2*whisker_x_bias} ${whisker_y} ${whisker_z}"
                rpy="0 0 0" />
      </xacro:whisker>
      <!-- whisker #7 -->
      <xacro:whisker name="${name}" id = "7" parent= "${name}_whisker_array_link" >
        <origin xyz="${whisker_x + 3*whisker_x_bias} ${whisker_y} ${whisker_z}"
                rpy="0 0 0" />
      </xacro:whisker>
      <!-- whisker #8 -->
      <xacro:whisker name="${name}" id = "8" parent= "${name}_whisker_array_link" >
        <origin xyz="${whisker_x + 4*whisker_x_bias} ${whisker_y} ${whisker_z}"
                rpy="0 0 0" />
      </xacro:whisker>
  </xacro:macro>
</robot>
